pkg/controlplane/controller/systemnamespaces/system_namespaces_controller.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package systemnamespaces
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	"k8s.io/klog/v2"
+)
+
+// Controller deletes expired apiserver leases.
+type Controller struct {
+	client kubernetes.Interface
+
+	namespaceLister listers.NamespaceLister
+	namespaceSynced cache.InformerSynced
+
+	systemNamespaces []string
+	interval         time.Duration
+}
+
+// NewController creates a new Controller to ensure system namespaces exist.
+func NewController(clientset kubernetes.Interface, namespaceInformer coreinformers.NamespaceInformer) *Controller {
+	systemNamespaces := []string{metav1.NamespaceSystem, metav1.NamespacePublic, v1.NamespaceNodeLease, metav1.NamespaceDefault}
+	interval := 1 * time.Minute
+
+	return &Controller{
+		client:           clientset,
+		namespaceLister:  namespaceInformer.Lister(),
+		namespaceSynced:  namespaceInformer.Informer().HasSynced,
+		systemNamespaces: systemNamespaces,
+		interval:         interval,
+	}
+}
+
+// Run starts one worker.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer klog.Infof("Shutting down system namespaces controller")
+
+	klog.Infof("Starting system namespaces controller")
+
+	if !cache.WaitForCacheSync(stopCh, c.namespaceSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.sync, c.interval, stopCh)
```

Comment:
Instead of doing that one per minute eventually like to take use of events, but let's not change it in this PR.

Reply:
I'll add it in a different commit so we can revert if necessary

Reply:
better different PR and beginning of next cycle, we are close to code freeze and I don't want to change the behavior now

Reply:
Yes - definitely.